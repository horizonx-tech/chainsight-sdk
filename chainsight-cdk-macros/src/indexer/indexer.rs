use proc_macro::TokenStream;
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    Ident, Result, Token,
};

pub struct Args {
    contract_name: String,
    event_name: String,
    args: Punctuated<NamedField, Token![,]>,
    address: String,
}
struct NamedField {
    name: Ident,
    _colon_token: Token![:],
    ty: syn::Type,
}
impl Parse for NamedField {
    fn parse(input: ParseStream) -> Result<Self> {
        Ok(NamedField {
            name: input.parse()?,
            _colon_token: input.parse()?,
            ty: input.parse()?,
        })
    }
}
impl Parse for Args {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let contract_name: syn::LitStr = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let event_name: syn::LitStr = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let args: Punctuated<NamedField, Token![,]> = Punctuated::parse_terminated(input)?;
        input.parse::<syn::Token![,]>()?;
        let address: syn::LitStr = input.parse()?;
        Ok(Args {
            contract_name: contract_name.value(),
            event_name: event_name.value(),
            args,
            address: address.value(),
        })
    }
}

pub fn gen(input: TokenStream) -> TokenStream {
    let Args {
        contract_name,
        event_name,
        args,
        address,
    } = syn::parse_macro_input!(input as Args);
    let struct_name = Ident::new(&format!("{}Event", event_name), event_name.span());

    quote! {}.into()
}
