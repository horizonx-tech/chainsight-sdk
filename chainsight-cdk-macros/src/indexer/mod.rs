pub mod indexer;
pub use indexer::*;
