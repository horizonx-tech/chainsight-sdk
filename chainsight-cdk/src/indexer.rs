use std::collections::HashMap;

use async_trait::async_trait;
use candid::CandidType;
use derive_more::{Display, From};

#[derive(Debug, Display, From)]
pub enum Error {}

pub trait Event: CandidType + Send + Sync + Clone + 'static {
    /// Create an event from a log.
    fn from<T>(log: T) -> Self;
}

#[async_trait]
pub trait Indexer {
    /// Get events from `from` to `to`. e.g. `from = 1`, `to = 10` will return events with id from 1 to 10.
    async fn get_from_to<U>(&self, from: u64, to: u64) -> Result<HashMap<u64, Vec<U>>, Error>;
    /// Save events to database.
    fn save<T>(&self, id: u64, elements: Vec<T>)
    where
        T: Event;
    /// Get the latest event id.
    fn get_last_number(&self) -> Result<u64, Error>;
    /// Get the event chunk size.
    fn get_event_chunk_size(&self) -> Result<u64, Error>;
    /// Set the event chunk size.
    fn set_event_chunk_size(&self, size: u64) -> Result<(), Error>;
    /// Get the last indexed event id.
    fn get_last_indexed(&self) -> Result<u64, Error>;
    /// Set the last indexed event id.
    fn set_last_indexed(&self, id: u64) -> Result<(), Error>;
    /// Index events.
    async fn index<T, U>(&self) -> Result<(), Error>
    where
        T: Event,
    {
        let last_indexed = self.get_last_indexed()?;
        let chunk_size = self.get_event_chunk_size()?;
        let from = last_indexed + 1;
        let to = last_indexed + chunk_size;
        self.get_from_to(from, to)
            .await?
            .into_iter()
            .map(|(k, v)| (k, v.into_iter().map(Event::from::<U>).collect::<Vec<T>>()))
            .for_each(|(k, v)| {
                self.save(k, v);
            });
        self.set_last_indexed(to)?;
        Ok(())
    }
}
