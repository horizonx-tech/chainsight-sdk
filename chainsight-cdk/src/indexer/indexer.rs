use std::collections::HashMap;

use async_trait::async_trait;
use candid::CandidType;
use derive_more::Display;
use serde::Deserialize;

use crate::storage::{self, Data};

#[derive(Debug, Display)]
pub enum Error {
    #[display(fmt = "Indexer error: {}", _0)]
    IndexerError(String),
    #[display(fmt = "Web3 error: {}", _0)]
    Web3Error(String),
    #[display(fmt = "Database error: {}", _0)]
    DatabaseError(String),
    #[display(fmt = "Other error: {}", _0)]
    OtherError(String),
}

#[derive(CandidType, Clone, Debug, Default, PartialEq, Deserialize)]
pub struct IndexingConfig {
    pub start_from: u64,
    pub chunk_size: Option<u64>,
}

pub trait Event<T>: CandidType + Send + Sync + Clone + 'static {
    fn from(event: T) -> Self
    where
        T: Into<Self>;
    fn tokenize(&self) -> Data;
    fn untokenize(data: Data) -> Self;
}

#[async_trait]
pub trait Finder<T>: Send + Sync + Clone + 'static {
    /// Get events from `from` to `to`. e.g. `from = 1`, `to = 10` will return events with id from 1 to 10.
    async fn find(&self, from: u64, to: u64) -> Result<HashMap<u64, Vec<T>>, Error>;
}

#[async_trait]
pub trait Indexer<T> {
    type Finder: Finder<T>;
    fn finder(&self) -> Self::Finder;
    /// Save events to database.
    fn save<E>(&self, id: u64, elements: Vec<E>)
    where
        E: Event<T>,
    {
        elements.iter().map(|e| e.tokenize()).for_each(|data| {
            storage::insert(id, data);
        });
    }
    fn between<E>(&self, from: u64, to: u64) -> Result<HashMap<u64, Vec<E>>, Error>
    where
        E: Event<T>,
    {
        let mut res = HashMap::new();
        storage::between(from, to).into_iter().for_each(|e| {
            let data =
                e.1.to_vec()
                    .iter()
                    .map(|e| Event::untokenize(e.clone()))
                    .collect::<Vec<E>>();
            res.insert(e.0, data);
        });
        Ok(res)
    }
    /// Get the last indexed event id.
    fn get_last_indexed(&self) -> Result<u64, Error> {
        Ok(storage::last(1).iter().map(|e| e.0).max().unwrap_or(0))
    }
    /// Index events.
    async fn index<E>(&self, cfg: IndexingConfig) -> Result<(), Error>
    where
        E: Event<T> + From<T>,
    {
        let last_indexed = self.get_last_indexed()?;
        let chunk_size = cfg.chunk_size.unwrap_or(500);
        let from = cfg.start_from.max(last_indexed + 1);
        let to = from + chunk_size;
        let elements: HashMap<u64, Vec<T>> = self.finder().find(from, to).await?;
        self.on_update::<E>(elements);
        Ok(())
    }

    fn on_update<E>(&self, elements: HashMap<u64, Vec<T>>)
    where
        E: Event<T> + From<T>,
    {
        elements
            .into_iter()
            .map(|(k, v)| (k, v.into_iter().map(Event::from).collect::<Vec<E>>()))
            .for_each(|(k, v)| {
                self.save(k, v);
            });
    }
}
