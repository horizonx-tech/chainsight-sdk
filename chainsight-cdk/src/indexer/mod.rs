mod indexer;
pub use indexer::*;
