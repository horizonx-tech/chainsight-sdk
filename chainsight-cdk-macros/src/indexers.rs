use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    parse_macro_input, Type,
};

pub struct Web3EventIndexerInput {
    out_type: syn::Type,
}

pub struct AlgorithmIndexerInput {
    in_type: syn::Type,
    call_method: syn::LitStr,
}

impl Parse for AlgorithmIndexerInput {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let in_ty: Type = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let call_method_str: syn::LitStr = input.parse()?;
        Ok(AlgorithmIndexerInput {
            in_type: in_ty,
            call_method: call_method_str,
        })
    }
}

impl Parse for Web3EventIndexerInput {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let out_type: Type = input.parse()?;
        Ok(Web3EventIndexerInput { out_type })
    }
}

pub fn web3_event_indexer(input: TokenStream) -> TokenStream {
    let Web3EventIndexerInput { out_type } = parse_macro_input!(input as Web3EventIndexerInput);
    let common = event_indexer_common(out_type.clone());
    quote! {
        #common
        fn indexer() -> chainsight_cdk::web3::Web3Indexer<#out_type> {
            chainsight_cdk::web3::Web3Indexer::new(get_logs, None)
        }
        #[ic_cdk::update]
        #[candid::candid_method(update)]
        async fn index() {
            indexer().index(get_config()).await.unwrap();
        }

        #[ic_cdk::query]
        #[candid::candid_method(query)]
        fn event_source() -> String {
            get_target_addr()
        }
    }
    .into()
}

pub fn algorithm_indexer(input: TokenStream) -> TokenStream {
    let AlgorithmIndexerInput {
        in_type,
        call_method,
    } = parse_macro_input!(input as AlgorithmIndexerInput);
    quote! {
        mod app;
        manage_single_state!("config", IndexingConfig, false);
        use chainsight_cdk::indexer::Indexer;
        fn indexer() -> chainsight_cdk::algorithm::AlgorithmIndexer<#in_type> {
            chainsight_cdk::algorithm::AlgorithmIndexer::new_with_method(proxy(), get_target(),stringify!(#call_method), app::persist)
        }
        #[ic_cdk::update]
        #[candid::candid_method(update)]
        async fn index() {
            let mut config = get_config();
            let stored = chainsight_cdk::storage::get_last_key();
            let stored_u64 = stored.parse::<u64>().unwrap_or(0);
            if stored_u64 > config.start_from {
                config.start_from = stored_u64;
            }
            indexer().index(config).await.unwrap();
        }
        fn get_target() -> candid::Principal {
            candid::Principal::from_text(get_target_addr()).unwrap()
        }

        #[ic_cdk::query]
        #[candid::candid_method(query)]
        fn event_source() -> candid::Principal {
            get_target()
        }

    }
    .into()
}

pub fn event_indexer_common(out_type: syn::Type) -> TokenStream2 {
    let output = quote! {
        manage_single_state!("config", IndexingConfig, false);
        #[ic_cdk::query]
        #[candid::candid_method(query)]
        pub fn events_from_to(from:u64, to: u64) -> HashMap<u64, Vec<#out_type>> {
            _events_from_to((from,to))
        }
        #[ic_cdk::query]
        #[candid::candid_method(query)]
        pub fn events_latest_n(n: u64) -> HashMap<u64, Vec<#out_type>> {
            let last_indexed = indexer().get_last_indexed().unwrap();
            _events_from_to((last_indexed - n, last_indexed))
        }
        fn _events_from_to(input: (u64,  u64)) -> HashMap<u64, Vec<#out_type>> {
            indexer().between(input.0,input.1).unwrap()
        }
        #[ic_cdk::query]
        #[candid::candid_method(query)]
        pub fn get_last_indexed() -> u64 {
            indexer().get_last_indexed().unwrap()
        }

        #[ic_cdk::update]
        #[candid::candid_method(update)]
        fn proxy_call(input: std::vec::Vec<u8>) -> std::vec::Vec<u8> {
            use chainsight_cdk::rpc::Receiver;
            chainsight_cdk::rpc::ReceiverProvider::<(u64, u64), HashMap<u64, Vec<#out_type>>>::new(
                proxy(),
                _events_from_to.clone(),
            )
            .reply(input)
        }
    };

    output
}
