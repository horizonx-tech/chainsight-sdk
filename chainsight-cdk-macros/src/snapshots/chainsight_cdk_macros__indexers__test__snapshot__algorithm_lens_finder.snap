---
source: chainsight-cdk-macros/src/indexers.rs
expression: formatted
---
pub async fn get_user(target_principal: String, args: u64) -> std::result::Result<User, String> {
    let method = finder_user(target_principal.clone());
    method
        .await
        .find(args)
        .await
        .map_err(|e| format!("{:?}", e))
}
pub async fn get_user_unwrap(target_principal: String, args: u64) -> User {
    finder_user(target_principal.clone())
        .await
        .find(args)
        .await
        .unwrap()
}
async fn finder_user(target_principal: String) -> chainsight_cdk::lens::AlgorithmLensFinder<User> {
    use chainsight_cdk::lens::LensFinder;
    let recipient = candid::Principal::from_text(target_principal).unwrap();
    chainsight_cdk::lens::AlgorithmLensFinder::new(chainsight_cdk::lens::LensTarget::<User>::new(
        _get_target_proxy(recipient).await,
        "get_user",
    ))
}

